README.md
# PaperTTY

## Partial refresh support for 4.2" *(2020-03-01)*

Thanks to @gdkrmr, the 4.2" module now has a new driver that supports partial refresh. Note that now the default behavior is to do partial updates and you'll need to add `--nopartial` to driver settings if you want to use the full refresh instead.

## Interactive Update *(2020-02-18)*

Starting `terminal` with the `--interactive` option adds a menu to the Ctrl-C handler that allows you to change the font, font size and spacing on the fly. This should be helpful when trying out fonts.

## Cursor Update *(2020-01-25)*

The `terminal` mode now supports different cursors. `--nocursor` had been deprecated in favor of `--cursor=none`. The usual underscore-like cursor is set via `--cursor=default`, which, as the name suggests, is also the default mode. There is also a `--cursor=block` option, which inverts the colors at the cursor's location. Finally, there is also an option to provide a numerical value, e.g. `--cursor=5`, which draws an underscore cursor shifted 5 pixels towards the top; `--cursor=default` and `--cursor=0` are thus equivalent.

## Unicode Update *(2020-01-22)*

On systems with /dev/vcsu* (e.g. Raspbian Stretch or Buster, with kernel 4.19+) and when using a TrueType font, the `terminal` mode now has full support for Unicode output. This is automatic, with a fallback to 8-bit if either of these requirements isn't met. We've also changed the standard encoding for 8-bit to ISO-8859-1, which should be a little closer to what the `vcs` device provides.

## IT8951 Optimization II *(2020-01-15)*

The previous optimization was improved further to speed up VNC output too. If you try it, please leave a comment at https://github.com/joukos/PaperTTY/issues/32 (and create a new issue if there's problems).

## IT8951 Optimization *(2020-01-11)*

An optimization by @chi-lambda for the IT8951 driver was merged and may speed up the bit-packing in TTY mode even by up to **10x** for these displays, so the refresh rates should be significantly better now. Please create an issue if there are problems (the unoptimized version can be checked out with the tag `v0.03_unoptimized`).

## IT8951 Update *(2019-11-03)*

Support for IT8951 based displays kindly provided by @fimad has been merged. This means that Waveshare 6", 7.8", 9.7" and 10.3" may work since they use this controller. At least the 6" and 9.7" displays have been verified to work. A big thanks to everyone involved with the discussion and code contributions (https://github.com/joukos/PaperTTY/issues/25)!

If you have one of these displays and want to test it, add your experiences to https://github.com/joukos/PaperTTY/issues/32.

Here is a picture of the 9.7" in action (by @math85360):

![](pics/vnc_97_inch_by_math85360.jpg)

[YouTube video](https://www.youtube.com/watch?v=J5WbhSV2E_U)

## VNC Update *(2019-07-07)*

So, it's been almost a year since last update and I've been very very busy.

Since I'm going to be even *busier* for the next two years or so, and probably won't have much time to spend on this project, I wanted to see if I could kludge this one last thing that I think is needed (and was referenced in https://github.com/joukos/PaperTTY/issues/23#issuecomment-435578128): a VNC client to PaperTTY.

**I'm happy to announce that it works!** It's not the most elegant or beautiful thing, but now you can run a graphical desktop and any programs you wish on these epaper screens. It also solves many problems and complexities with text-only terminals, namely encoding/font issues and other quirks. I've tested it with the 4.2" and the 2.13" displays.

The performance is pretty much the same as before, maybe a tad more image processing though, but the bottleneck is the display refresh anyway. All the complex stuff is done by [vncdotool](https://github.com/sibson/vncdotool) by Marc Sibson, since it was the most mature-seeming VNC library I could find that would work with Python 3 without extra work. It's also overkill for the job - all that's needed is to read the screen content periodically.

The way this works is pretty simple: PaperTTY will connect to the VNC server, then just reads the screen content in a loop. If the image changes, the region containing the changes will be updated on the display. This is not actually very efficient and originally I planned to have it update the native VNC "dirty rectangles", but it was simpler to kludge it like this for a start, and it seems to work fine, so it's perhaps easier to make better now.

### Some benefits and features

- **Run all CLI *and* GUI programs**
- Dithers colors so you can throw anything at it and hopefully it's legible on the display
- About as fast as just using the terminal mode, but much less hassle (sort of)
- Has options for rotating and inverting the colors
- Rescales the VNC screen to fit the display (panning not implemented)
- Simplistic, surely buggy and doesn't fix the other problems PaperTTY might have

### Quick start

- Install PaperTTY as usual
   - `requirements.txt` is updated to include `vncdotool` - run `pip install -r requirements.txt` if you have an existing virtualenv
- Start a VNC server somewhere (on the RPi for example)
   - ie. `vncserver -geometry 250x128 :1`
- Run PaperTTY as usual, but use the `vnc` subcommand (see `--help` for options)
   - ie. `sudo ~/.virtualenvs/papertty/bin/python3 ./papertty.py --driver epd2in13 vnc --display 1 --password supAPass --sleep 0.1 --rotate 90`
   - This would (by default) connect to `localhost`, display `1` (= port 5901), using the 2.13" driver, specifies the password, sleeps 0.1 seconds after each update, and rotates the screen by 90 degrees
- If image looks wonky, make sure you have the right orientation for the VNC server (ie. `-geometry 300x400` vs. `-geometry 400x300`) and the correct `--rotate` value.

### Standalone RPi with mouse and keyboard

So, you have your nice display and a RPi that you want to combine into a cool e-ink laptop, using peripherals plugged into the RPi. There's currently no automation or service files for this purpose, but what you can try is this (for the 2.13"):

- Install or enable a desktop environment / window manager, unless it already exists (ie. if you plug in the HDMI port, you should get a graphical session)
- Install `x11vnc` on the RPi
- Plug in keyboard and mouse
- Start a `tmux` session on the RPi and run the following commands, in their own windows (to easily find out if something goes wrong)
- ```sh
  # tmux window/pane #0
  sudo lightdm    # start lightdm if not already started
  export $(sudo x11vnc -findauth)	# find the X authority file (probably /var/run/lightdm/root/:0)
  sudo x11vnc -usepw -display :0 -forever	# start x11vnc
  # tmux window/pane #1
  sudo /home/pi/.virtualenvs/papertty/bin/python papertty.py --driver epd2in13 scrub      # scrub for clear image
  sudo /home/pi/.virtualenvs/papertty/bin/python papertty.py --driver epd2in13 vnc --display 0 --password supAPass --sleep 0.1 --rotate 90   # display the session
  ```
- If all went well, you should now be able to see the login screen (or whatever is in your display :0) and can interact with the GUI using mouse and keyboard plugged into the RPi, making it a cool standalone unit.

**Here's a quick video showing this in action**

[![YouTube Video](https://img.youtube.com/vi/Kpohs9iqUBk/0.jpg)](https://www.youtube.com/watch?v=Kpohs9iqUBk)

The RPi Zero W is quite sluggish though and the default resolution was used here, so downscaling it to 2.13" makes it pretty unreadable even without the background images and other crud. It's a proof of concept anyway and just needs a bit of polish.

I'll try to add more details later if I have time, but here's some more screenshots:

**FreeCiv, downscaled from 800x600**

![](pics/vnc_freeciv.jpg)

**IRC and surfing action**

![](pics/vnc_irc.jpg)

**Freedoom, why not**

![](pics/vnc_freedoom.jpg)

**Early image of testing some X progs**

![](pics/vnc.jpg)

**And a YouTube video to showcase some partial refresh action**

[![YouTube Video](https://img.youtube.com/vi/3o9Z2Ujdr9Q/0.jpg)](https://www.youtube.com/watch?v=3o9Z2Ujdr9Q)

As always, create issues if there's problems. I'd also be very interested if someone with a 9.7" display gets this to work with partial refresh.

*The rest of this page is not updated to cover the VNC feature yet.*

--- 

![](pics/logo.jpg)

This is an experimental command-line driven Python module to render the contents of a Linux virtual terminal (`/dev/tty[1-63]`) **or standard input** onto a [Waveshare](https://www.waveshare.com/) e-Paper display. See [list of supported displays](drivers/README.md).

*Note: Testing has been minimal and I probably forgot something, so 'caveat utilitor'. I am also not affiliated with Waveshare in any way.*

*Note: relevant changes listed in the [changelog](CHANGELOG.md).*

#### Some features

- Designed to be used with a **Raspberry Pi** and Raspbian.
- It should enable you to run interactive console programs (`vim`, `tmux`, `irssi`, `nethack` ...) and display whatever you want easily with scripts.
- Especially with a small font, it is fast enough for interactive use but could be improved to be [even faster](http://benkrasnow.blogspot.com/2017/10/fast-partial-refresh-on-42-e-paper.html). Also, it's quite a bit snappier on the Raspberry Pi 3 than the Zero.
- Only the changed region is updated on the display, so typing is faster than full screen scrolling.
- The cursor is also drawn and the image updated as it moves.
- Flicker-free.
- Allows changing the font, font size, orientation and some other parameters.
- Supports TrueType and bitmap fonts (in PIL format).
- Bundled with a `systemd` service unit to start the service early at boot and gracefully stop it.

**It isn't perfect and has only been tested with the monochrome 2.13" HAT, but it *might* work for other models too, and allows you to at least *try*.**

- *The PaperTTY code is in the public domain and you run it **at your own risk.***
- The driver code (in `drivers/`) is **GPL 3.0** licensed, because it is based on Waveshare's GPL code - you still run it **at your own risk.**

## Screenshots

**Collage of running various programs in `tmux`**

![](pics/collage.png)

**Running Nethack outside in the noon sun, powered directly by a solar panel, connected to a Bluetooth keyboard**

![](pics/sunlight.jpg)

**Action video - terminal usage (Raspberry Pi Zero W)**

Showcasing input feedback.

[![Youtube Video](https://img.youtube.com/vi/mXBS4l3OvyE/0.jpg)](https://www.youtube.com/watch?v=mXBS4l3OvyE)

**Action video 2 - cacafire (Raspberry Pi 3)**

The RPi3 is noticeably faster - `cacafire` is 3x slower on the Zero. Typical terminal usage works pretty well.

[![Youtube Video](https://img.youtube.com/vi/yWpT0xk8ufY/0.jpg)](https://www.youtube.com/watch?v=yWpT0xk8ufY)

## Installation

All of the code was written for Raspbian Stretch and Python 3.5+. These instructions assume you're going to run this on a Raspberry Pi, otherwise you're on your own.

The code includes a reimplementation/refactoring of the Waveshare [reference drivers](https://github.com/soonuse/epd-library-python) - unlike the rest of the code which is CC0, **the drivers have the GPL 3.0 license**, because that's what Waveshare used. The drivers for models that aren't in the repo have been acquired from their Wiki's demo code packages.

See the [driver page](drivers/) for details and the supported models.

*The earlier, initial version of PaperTTY (tag: `v0.01`) did not have instructions for using virtualenv (though it would work) - you can still run it as before using the system packages and alongside this new version. Using the virtualenv means that PIL and Pillow can also coexist on the same system.*

#### Requirements

- Enable SPI (`sudo raspi-config`)
  - `Interfacing Options -> SPI -> Yes`
  - Reboot

#### Steps

1. **Clone the repo somewhere and enter the directory**
   - `git clone https://github.com/joukos/PaperTTY.git`
   - `cd PaperTTY`
2. **Install virtualenv and libopenjp2**
   - `sudo apt install virtualenvwrapper python3-virtualenv libopenjp2-7`
3. **Source the wrapper to use `mkvirtualenv` (*you may want to add this to `~/.bashrc`*)**
   - `source /usr/share/virtualenvwrapper/virtualenvwrapper.sh`
4. **Create the Python 3 virtualenv and install packages in `requirements.txt`**
   - `mkvirtualenv -p /usr/bin/python3 -r requirements.txt papertty`
   - This will create `~/.virtualenvs/papertty` which contains the required environment
5. **After creating the virtualenv, it should become active and you should see `(papertty)` on your prompt**
   - **Note:** the software needs to be run with `sudo` in the typical case, *so you need to explicitly start the interpreter within the virtualenv - otherwise the program attempts to import system packages instead*
   - **You should now be able to run `sudo ~/.virtualenvs/papertty/bin/python3 ./papertty.py list` to see the available drivers and start using the software**
6. **Not really needed, but to (de)activate the virtualenv afterwards, run:**
   -  `~/.virtualenvs/papertty/bin/activate` - activate the virtualenv
      -  Or, `workon papertty` if you have sourced `virtualenvwrapper.sh`
   -  `deactivate` - deactivate the virtualenv

#### Alternative install without virtualenv, using system packages

- If you don't care to use the virtualenv, just install the requirements as system packages:
  - `sudo apt install python3-rpi.gpio python3-spidev python3-pil python3-click`
  - And run the program directly: `sudo ./papertty.py list`

## Fonts

You can use TrueType fonts or bitmap fonts, but the bitmap fonts need to be in the right format. With bitmap fonts the `--size` option is ignored.

Included as default is a very small bitmap font called [Tom Thumb](https://robey.lag.net/2010/01/23/tiny-monospace-font.html), it is fairly readable for its tiny size and fits 20 rows with 62 columns on the 2.13". Thanks go to Brian Swetland and Robey Pointer for their work on the font and for releasing it under [CC0](https://creativecommons.org/publicdomain/zero/1.0/legalcode).

Another included font is the [nanofont](https://github.com/Michaelangel007/nanofont3x4), which is an extremely tiny (3x4 pixels) font and also released under CC0. Thanks go to the author, Michael Pohoreski. The conversion was done by generating the BMP, then transformed it with Pillow so that everything was on one line, then used [Fony](http://hukka.ncn.fi/?fony) to save a BDF and converted that to PIL.

Why would you use such a microscopic font, I hear you ask? One good reason is that some programs refuse to start unless the terminal size is big enough, and using this font will allow you to get things *theoretically* readable and run those programs even on the smaller displays. One example being **Dungeon Crawl Stone Soup** which wouldn't otherwise start on the 2.13" display (*hooray!*):

![](pics/dcss.jpg)

Playing the game like this would be quite challenging, however...

Unless you're happy with the awesome default font, find a nice *monospaced* TrueType or bitmap font: Andale Mono (`sudo apt install ttf-mscorefonts-installer`) is pretty great for very small sizes and on the 2.13" (128x250 pixels) can fit 17 rows and 50 columns

- You *can* use a proportional font but the terminal will probably look horrible

Pillow includes a utility called `pilfont.py`, you can use this to convert a BDF/PCF font file into a `.pil` and a `.pbm` (I didn't have luck with some fonts - remember to use the `pilfont.py` version that's on your Pi):

```
# convert Terminus 
gunzip -c /usr/share/fonts/X11/misc/ter-u12b_unicode.pcf.gz > terminus-12.pcf
pilfont.py terminus-12.pcf
# you should get terminus-12.pil that you can pass with the --font option
```

![](pics/terminus.jpg)

All font options expect a path to the font file - the system font directories are not searched for them.

## Usage

**If you have the requirements installed in a virtualenv, remember to use its interpreter when running the program with sudo or being root:**  `sudo ~/.virtualenvs/papertty/bin/python3 ./papertty.py` 

- You'll want to `sudo` unless you've set it up so that SPI works without and you've given read access to `/dev/vcsa*`

To do anything useful, you'll need to tell the script which model you're using - in my case this would be **`epd2in13`**. Use the top-level option **`--driver`** to set the desired driver.

Append `--help` with the subcommands to get help with their parameters.

You can just edit `papertty.py` to your liking - the code is very simple and commented.

**Top-level options**

These should come before the subcommands and they control general settings.

| Option          | Description                                              | Default      |
| --------------- | -------------------------------------------------------- | ------------ |
| `--driver NAME` | Select driver to use - **required**                      | *no default* |
| `--nopartial`   | Disable partial refresh even if the display supported it | disabled     |
| `--encoding NAME` | Select encoding to use                                 | `utf-8`      |

**Note:** The encoding settings are a bit questionable right now - encoding/decoding is done explicitly to have `ignore` on any errors, but I think this needs some more work as it's not an entirely trivial issue. If you feel like there's a big dum-dum in the code regarding these, a PR is *very appreciated*.

**Note 2:** *To get scandinavian accents to show (`ä`,`ö` etc.), try `--encoding cp852`.*

### Commands

#### `list` - List display drivers

```sh
# Example
sudo ./papertty.py list
```

#### `scrub` - Scrub display

*This command mostly makes sense with the partial refresh models, although you can run it with full refresh too - it's just going to take a pretty long time to run. I needed this because my own unit can't handle a full refresh so it was the only way to clear the screen properly!*

If you're left with "burn-in" or the display doesn't seem to work properly, this usually helps to even it out (may even need to run it twice sometimes if the display is not in a steady state).

This will slowly fill the screen with bands of black, then white.

Option | Description | Default
---    | --- | ---
`--size N` | Chunk width (pixels) to fill with (valid values: `8-32`) | `16` 

```sh
# Example
sudo ./papertty.py --driver epd2in13 scrub
```

#### `stdin` - Render standard input

Render `stdin` on the display, simple as that. Leaves the image on the display until something else overwrites it. Very useful for showing script output or just about anything that updates irregularly.

Option | Description | Default
---    | --- | ---
`--font FILENAME` | Path to a TrueType or PIL font to use - **strongly recommended to use monospaced** | `tom-thumb.pil`
`--size N` | Font size | `8` 
`--width N` | Fit to a particular width (characters) | display width / font width
`--portrait` | Enable portrait mode | disabled
`--nofold` | Disable folding (ie. don't wrap to width) | disabled
`--spacing` | Set line spacing | `0` 


```sh
# Example
cowsay "Hello World" | sudo ./papertty.py --driver epd2in13 stdin --nofold
```

#### `terminal` - Render a virtual terminal

The most prominent feature.

This requires read permission to the virtual console device (`/dev/vcsa[1-63]`) and optionally write permission to the associated terminal device (`/dev/tty[1-63]`) if you want to set the TTY size via `ioctl`s.

If you're going to use `terminal` with a display that doesn't support partial refresh, you probably want to set `--sleep` a bit larger than the default, such as a few seconds, unless you enjoy blinking.

**The process handles two signals:**

- **`SIGINT`** - stop and clear the screen (unless `--noclear` was given), same as pressing Ctrl-C
    - `sudo pkill -INT -f papertty.py`
    - By default, the `systemd` service unit attempts to stop the process using SIGINT
- **`SIGUSR1`** - apply scrub and keep running
    - `sudo pkill -USR1 -f papertty.py`

See details on how all of this works further down this document.

Option | Description | Default
---    | --- | ---
`--vcsa FILENAME` | Virtual console device (`/dev/vcsa[1-63]`) | `/dev/vcsa1`
`--font FILENAME` | Path to a TrueType or PIL font to use - **strongly recommended to use monospaced** | `tom-thumb.pil`
`--size N` | Font size | `8` 
`--noclear` | Leave display content on exit | disabled
`--nocursor` | Don't draw cursor | disabled
`--sleep` | Minimum delay between screen updates (seconds) | `0.1` 
`--rows` | Set TTY rows (`--cols` required too) | *no default*
`--cols` | Set TTY columns (`--rows` required too) | *no default*
`--portrait` | Enable portrait mode | disabled
`--flipx` | Mirror X axis (experimental / buggy) | disabled
`--flipy` | Mirror Y axis (experimental / buggy) | disabled
`--spacing` | Set line spacing | `0` 
`--scrub` | Apply scrub when starting | disabled
`--autofit` | Try to automatically set terminal rows/cols for the font | disabled


```sh
# Examples

# by default the first virtual terminal (/dev/vcsa1 == /dev/tty1) is displayed
sudo ./papertty.py --driver epd2in13 terminal

# set font size to 16, update every 10 seconds, set terminal rows/cols to 10x20
sudo ./papertty.py --driver epd2in13 terminal --size 16 --sleep 10 --rows 10 --cols 20

# auto-fit terminal rows/cols for the font and use a bitmap font
# (fitting may not work for very small fonts in portrait mode because of terminal restrictions)
sudo ./papertty.py --driver epd2in13 terminal --autofit --font myfont.pil
```

## How to use the terminal

#### Logging in?

After you've gotten the terminal to render, you'll want to run something there.

As the program mirrors the system virtual terminals, you can either attach a keyboard to the Pi and simply log in **or** use the `openvt` program to start something there without messing around with cables, if you already have SSH access.

**The following commands are run over SSH.**

For example, to start `htop` for user `pi` on `tty1` (via `sudo`, *twice*):

```sh
# "as a sudoer, start sudo forcibly on VT 1 (tty1) to run 'htop' as the user 'pi'"
sudo openvt -fc 1 -- sudo -u pi htop
```

After you exit the process, `agetty` may go haywire though (hogging CPU). Give it a nudge to fix it:

```sh
sudo pkill agetty
```

And you should have the login prompt there again.

In practice, you'll want to use **`tmux`** (or **`screen`**, if you prefer) to have the most flexible control over the terminal (these are terminal multiplexers, and if you haven't used one before, now is the time to start):

```sh
# start a new tmux session (or just run 'tmux' with a connected keyboard)
sudo openvt -fc 1 -- sudo -u pi tmux
# (see the session starting up on the display)
# now, attach to the session
tmux attach
```

Lo and behold! You should now be attached to the tiny session visible on the display.

You can kill the `papertty.py` process at any time - the stuff that runs in the TTY will be unaffected (unless they react badly to console resizing) and you can just restart the `terminal` to get the display back and play around with the settings.

#### Start up at boot

A simple `systemd` service unit file is included with the package, called `papertty.service`. It calls `start.sh` so that instead of editing the service file, you can edit the start script (and easily add whatever you need) without needing to run `systemctl daemon-reload` all the time.
- You can simply put the command in the service file too, it's your choice
- You probably want to set the script to be owned and writable by root only: `sudo chown root:root start.sh; sudo chmod 700 start.sh`
- **Remember: to run the command under the virtualenv, you need to run the `python3` command from within the virtualenv's `bin` directory - this will ensure the environment is correct**

To have the display turn on at boot, first **edit** the command you're happy with into `start.sh`:

```sh
# Remember: you probably want to set rows and cols here, because at reboot they're reset.
# Also, when booting up after a power cycle the display may have some artifacts on it, so 
# you may want to add --scrub to get a clean display (during boot it's a bit slower than usual)
VENV="/home/pi/.virtualenvs/papertty/bin/python3"
${VENV} papertty.py --driver epd2in13 terminal --autofit
```

Then make sure you have the right paths set in the service file:

```sh
...
### Change the paths below to match yours
WorkingDirectory=/home/pi/code/PaperTTY
ExecStart=/home/pi/code/PaperTTY/start.sh
###
...
```

Then (read the unit file more carefully and) do the following steps:

```sh

CHANGELOG.md
# Changelog

- **2020-03-01**
  - Added partial refresh support for 4.2".
- **2020-02-18**
  - Added interactive mode.
- **2020-01-25**
  - Added support for more cursors.
- **2020-01-22**
  - Added Unicode support for systems providing /dev/vcsu
- **2019-12-07**
  - Added basic support for EPD7in5v2 (GDEW075T7)
- **2019-11-03**
  - Added IT8951 support
- **2019-07-07**
  - Added VNC / GUI support
- **2018-08-31**
  - Installation instructions had an issue: running with `sudo` didn't respect the virtualenv - fixed
- **2018-08-26**
  - Added [nanofont](https://github.com/Michaelangel007/nanofont3x4) in PIL format
- **2018-08-25**
  - Now **all** the SPI models listed in Waveshare's Wiki have a driver implemented - all it needs now is some testing to see if any of it works
  - Fixed little issue with `--nopartial`: the screen was updated twice instead of once when enabled
  - Added support for:
    - **EPD 2.13" D (monochrome, flexible)**
      - Note: the original driver for this model seemed to have been written by someone who didn't write the rest of the drivers - it appeared to be essentially broken and written while drunk, I tried to fix it but who knows if it'll work
  - Added support for:
    - **EPD 5.83" (monochrome)**
    - **EPD 5.83" B (black/white/red)**
    - **EPD 5.83" C (black/white/yellow)**
  - Just one still missing:
    - **~~EPD 2.13" D (monochrome, flexible)~~**
- **2018-08-24**
  - **Major** overhaul
    - **Please create an issue (or pull request) if things don't work**
    - Converted code to Python 3 only
    - Install by default into a virtualenv
      - Can co-exist with the previous version
    - Replace old PIL with Pillow
    - **Bundled reorganized/modified display drivers to (*hopefully*) support more displays**
      - Created class structure to reduce code duplication in the drivers (got rid of ~60%)
        - Perhaps overkill and more complicated, but meh, I just couldn't bear the repetition and accommodating all the models would have been messy otherwise
      - **Supported models (SPI)**
        - **EPD 1.54" (monochrome) - [probably works, with partial refresh]**
        - **EPD 1.54" B (black/white/red)**
        - **EPD 1.54" C (black/white/yellow)**
        - **EPD 2.13" (monochrome) - [TESTED, with partial refresh]** 
        - **EPD 2.13" B (black/white/red)**
        - **EPD 2.13" C (black/white/yellow)** - should work with `EPD2in13b`
        - **EPD 2.7" (monochrome)**
        - **EPD 2.7" B (black/white/red)**
        - **EPD 2.9" (monochrome) - [probably works, with partial refresh]**
        - **EPD 2.9" B (black/white/red)**
        - **EPD 2.9" C (black/white/yellow)** - should work with `EPD2in9b`
        - **EPD 4.2" (monochrome)**
        - **EPD 4.2" B (black/white/red)**
        - **EPD 4.2" C (black/white/yellow)** - should work with `EPD4in2b`
        - **EPD 7.5" (monochrome)**
        - **EPD 7.5" B (black/white/red)**
        - **EPD 7.5" C (black/white/yellow)** - should work with `EPD7in5b`
      - **Missing models**
        - **EPD 2.13" D (monochrome, flexible)**
        - **~~EPD 5.83" (monochrome)~~**
        - **~~EPD 5.83" B (black/white/red)~~**
        - **~~EPD 5.83" C (black/white/yellow)~~**
      - **Special drivers**
        - **Dummy - no-op driver**
        - **Bitmap - output frames as bitmap files (for debugging)**
      - *Note: PaperTTY doesn't use the red/yellow colors ... yet*
      - See `drivers/README.md`
    - Some CLI changes related to driver selection
    - I learned that my particular unit has some flaw that means it doesn't do full refreshes properly (never has)

      - ***Doh!***
      - **I'll just assume it works as expected with other people's units**
    - Added new video
    - Heard that a partial refresh LUT for the 7.5" is nontrivial to do if at all possible, so best not to get your hopes up too much regarding those - there will probably be better panels available eventually
- **2018-08-16**
  - Included a very tiny bitmap font, "Tom Thumb", to use as default
- **2018-08-14**
  - Added support for PIL bitmap fonts (autodetected)
- **2018-08-13**
  - After browsing the Waveshare Wiki a bit, it seems that the smaller models support partial refresh out of the box but the bigger ones need some (hopefully minor) modification, if I ever get one I'll look into it
    - My guess is that this code should work as-is on the **Black/White 1.54", 2.13" and 2.9"** models - no guarantees though
    - The **2.7", 4.2", 5.83", 7.5"** models have slightly different code and need LUT modification or some hacking to achieve partial refresh, and I'm not sure if it's feasible to get it to work with the color models at all
    - ~~Modifying the code to work with full refreshes should be pretty easy if you happen to own one of the bigger monochrome displays~~
      - *This is now done, and may work with color ones as well (2018-08-24)*


drivers/drivers_base.py
#     Copyright (c) 2018 Jouko Strömmer
#     Copyright (c) 2017 Waveshare
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.

from abc import ABC, abstractmethod
from PIL import Image

import time

# if rpi libs are not found, don't care - hope that we don't end up
# using them
try:
    import spidev
    import RPi.GPIO as GPIO
except ImportError:
    pass


class DisplayDriver(ABC):
    """Abstract base class for a display driver - be it Waveshare e-Paper, PaPiRus, OLED..."""

    # override these if needed
    white = 255
    black = 0

    def __init__(self):
        super().__init__()
        self.name = None
        self.width = None
        self.height = None
        self.colors = None
        self.type = None
        self.supports_partial = None
        self.partial_refresh = None

    @abstractmethod
    def init(self, **kwargs):
        """Initialize the display"""
        pass

    @abstractmethod
    def draw(self, x, y, image):
        """Draw an image object on the display at (x,y)"""
        pass

    def scrub(self, fillsize=16):
        """Scrub display - only works properly with partial refresh"""
        self.fill(self.black, fillsize=fillsize)
        self.fill(self.white, fillsize=fillsize)

    def fill(self, color, fillsize):
        """Slow fill routine"""
        image = Image.new('1', (fillsize, self.height), color)
        for x in range(0, self.height, fillsize):
            self.draw(x, 0, image)

    def clear(self):
        """Clears the display"""
        image = Image.new('1', (self.height, self.width), self.black)
        self.draw(0, 0, image)
        image = Image.new('1', (self.height, self.width), self.white)
        self.draw(0, 0, image)

class SpecialDriver(DisplayDriver):
    """Drivers that don't control hardware"""
    default_width = 640
    default_height = 384

    def __init__(self, name, width, height):
        super().__init__()
        self.name = name
        self.width = width
        self.height = height
        self.type = 'Dummy display driver'

    @abstractmethod
    def init(self, **kwargs):
        pass

    @abstractmethod
    def draw(self, x, y, image):
        pass

    def scrub(self, fillsize=16):
        pass


class Dummy(SpecialDriver):
    """Dummy display driver - does not do anything"""

    def __init__(self):
        super().__init__(name='No-op driver', width=self.default_width, height=self.default_height)

    def init(self, **kwargs):
        pass

    def draw(self, x, y, image):
        pass


class Bitmap(SpecialDriver):
    """Output a bitmap for each frame - overwrite old ones"""

    def __init__(self, maxfiles=5, file_format="png"):
        super().__init__(name="Bitmap output driver", width=self.default_width, height=self.default_height, )
        self.maxfiles = maxfiles
        self.current_frame = 0
        self.frame_buffer = None
        self.file_format = file_format

    def init(self, **kwargs):
        self.frame_buffer = Image.new('1', (self.width, self.height), 255)
        self.current_frame = 0

    def draw(self, x, y, image):
        self.frame_buffer.paste(image, box=(x, y))
        self.frame_buffer.save("bitmap_frame_{}.{}".format(self.current_frame, self.file_format))
        self.current_frame = (self.current_frame + 1) % self.maxfiles


class WaveshareEPD(DisplayDriver):
    """Base class for Waveshare displays with common code for all - the 'epdif.py'
    - 1.54" , 1.54" B , 1.54" C
    - 2.13" , 2.13" B
    - 2.7"  , 2.7" B
    - 2.9"  , 2.9" B
    - 4.2"  , 4.2" B
    - 7.5"  , 7.5" B
    """

    # Common commands
    GET_STATUS = 0x71

    # These pins are common across all models
    RST_PIN = 17
    DC_PIN = 25
    CS_PIN = 8
    BUSY_PIN = 24

    # Some models implement rotation in their code
    ROTATE_0 = 0x00
    ROTATE_90 = 0x01
    ROTATE_180 = 0x02
    ROTATE_270 = 0x03

    # SPI device, bus = 0, device = 0

    # SPI methods

    @staticmethod
    def epd_digital_write(pin, value):
        GPIO.output(pin, value)

    @staticmethod
    def epd_digital_read(pin):
        return GPIO.input(pin)

    @staticmethod
    def epd_delay_ms(delaytime):
        time.sleep(float(delaytime) / 1000.0)

    def spi_transfer(self, data):
        self.SPI.writebytes(data)

    def epd_init(self):
        GPIO.setmode(GPIO.BCM)
        GPIO.setwarnings(False)
        GPIO.setup(self.RST_PIN, GPIO.OUT)
        GPIO.setup(self.DC_PIN, GPIO.OUT)
        GPIO.setup(self.CS_PIN, GPIO.OUT)
        GPIO.setup(self.BUSY_PIN, GPIO.IN)
        self.SPI = spidev.SpiDev(0, 0)
        self.SPI.max_speed_hz = 2000000
        self.SPI.mode = 0b00
        return 0

    # Basic functionality

    def __init__(self, name, width, height):
        super().__init__()
        self.name = name
        self.width = width
        self.height = height
        self.type = 'Waveshare e-Paper'

    def digital_write(self, pin, value):
        self.epd_digital_write(pin, value)

    def digital_read(self, pin):
        return self.epd_digital_read(pin)

    def delay_ms(self, delaytime):
        self.epd_delay_ms(delaytime)

    def send_command(self, command):
        self.digital_write(self.DC_PIN, GPIO.LOW)
        # the parameter type is list but not int
        # so use [command] instead of command
        self.spi_transfer([command])

    def send_data(self, data):
        self.digital_write(self.DC_PIN, GPIO.HIGH)
        # the parameter type is list but not int
        # so use [data] instead of data
        self.spi_transfer([data])

    def reset(self):
        self.digital_write(self.RST_PIN, GPIO.LOW)
        self.delay_ms(200)
        self.digital_write(self.RST_PIN, GPIO.HIGH)
        self.delay_ms(200)

    def init(self, **kwargs):
        pass

    def draw(self, x, y, image):
        pass
